src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs

```diff
@@ -154,6 +154,312 @@ interface I1
+            // PROTOTYPE: should warn about copying 'this' when calling M2
+            verifier.VerifyDiagnostics();
+            verifier.VerifyIL("S.M1", @"
+{
+  // Code size       51 (0x33)
+  .maxstack  1
+  .locals init (S V_0, //copy
+                S V_1)
+  IL_0000:  ldarg.0
+  IL_0001:  ldobj      ""S""
+  IL_0006:  stloc.1
+  IL_0007:  ldloca.s   V_1
+  IL_0009:  call       ""void S.M2()""
+  IL_000e:  ldarg.0
+  IL_000f:  ldfld      ""int S.i""
+  IL_0014:  call       ""void System.Console.Write(int)""
+  IL_0019:  ldarg.0
+  IL_001a:  ldobj      ""S""
+  IL_001f:  stloc.0
+  IL_0020:  ldloca.s   V_0
+  IL_0022:  call       ""void S.M2()""
+  IL_0027:  ldloc.0
+  IL_0028:  ldfld      ""int S.i""
+  IL_002d:  call       ""void System.Console.Write(int)""
+  IL_0032:  ret
+}");
+        }
+
+        [Fact]
+        public void ReadOnlyAccessor_CallNormalMethod()
+        {
+            var csharp = @"
+public struct S
+{
+    public int i;
+
+    public int P
+    {
+        readonly get
+        {
+            // should create local copy
+            M();
+            System.Console.Write(i);
+
+            // explicit local copy, no warning
+            var copy = this;
+            copy.M();
+            System.Console.Write(copy.i);
+
+            return i;
+        }
+    }
+
+    void M()
+    {
+        i = 23;
+    }
+
+    static void Main()
+    {
+        var s = new S { i = 1 };
+        _ = s.P;
+    }
+}
+";
+
+            var verifier = CompileAndVerify(csharp, expectedOutput: "123");
+            // PROTOTYPE: should warn about copying 'this' when calling M
+            verifier.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ReadOnlyStruct_CallNormalMethodOnField()
+        {
+            var csharp = @"
+public readonly struct S1
+{
+    public readonly S2 s2;
+    public void M1()
+    {
+        s2.M2();
+    }
+}
+
+public struct S2
+{
+    public int i;
+    public void M2()
+    {
+        i = 42;
+    }
+
+    static void Main()
+    {
+        var s1 = new S1();
+        s1.M1();
+        System.Console.Write(s1.s2.i);
+    }
+}
+";
+            CompileAndVerify(csharp, expectedOutput: "0");
+        }
+
+        [Fact]
+        public void ReadOnlyMethod_CallNormalMethodOnField()
+        {
+            var csharp = @"
+public struct S1
+{
+    public S2 s2;
+    public readonly void M1()
+    {
+        // warn on local copy
+        s2.M2();
+        System.Console.Write(s2.i);
+
+        // no warning on explicit copy
+        var copy = s2;
+        copy.M2();
+        System.Console.Write(copy.i);
+    }
+}
+
+public struct S2
+{
+    public int i;
+    public void M2()
+    {
+        i = 23;
+    }
+
+    static void Main()
+    {
+        var s1 = new S1() { s2 = new S2 { i = 1 } };
+        s1.M1();
+    }
+}
+";
+            var verifier = CompileAndVerify(csharp, expectedOutput: "123");
+            // PROTOTYPE: should warn about calling M2
```

Comment:
Note that this is still under discussion in LDM. We may end up dropping this warning.